src/chunkserver/datastore/file_pool.cpp

```diff
@@ -244,24 +252,197 @@ bool FilePool::CheckValid() {
+    pushBack(cleanChunks_, chunkid, currentState_.cleanChunksLeft);
+    return true;
+}
+
+void FilePool::CleanWorker() {
+    auto sleepInterval = kSuccessSleepMsec_;
+    while (cleanSleeper_.wait_for(sleepInterval)) {
+        sleepInterval = CleaningChunk() ? kSuccessSleepMsec_ : kFailSleepMsec_;
+    }
+}
+
+bool FilePool::StartCleaning() {
+    if (poolOpt_.needClean && !cleanAlived_.exchange(true)) {
+        ReadWriteThrottleParams params;
+        params.iopsTotal = ThrottleParams(poolOpt_.iops4clean, 0, 0);
+        cleanThrottle_.UpdateThrottleParams(params);
+
+        cleanThread_ = Thread(&FilePool::CleanWorker, this);
+        LOG(INFO) << "Start clean thread ok.";
+    }
+
+    return true;
+}
+
+bool FilePool::StopCleaning() {
+    if (cleanAlived_.exchange(false)) {
+        LOG(INFO) << "Stop cleaning...";
+        cleanSleeper_.interrupt();
+        cleanThread_.join();
+        LOG(INFO) << "Stop clean thread ok.";
+    }
+
+    return true;
+}
+
+uint64_t FilePool::GetChunk(bool needClean, bool* isCleaned) {
+    auto smartPop = [this](std::vector<uint64_t>& chunks1,
+                           std::vector<uint64_t>& chunks2,
+                           uint64_t& chunksLeft1,
+                           uint64_t& chunksLeft2,
+                           int& from) -> uint64_t {
+        std::unique_lock<std::mutex> lk(mtx_);
+        from = 0;
+        uint64_t chunkid = 0;
+        if (chunks1.size() > 0) {
+            chunkid = chunks1.back();
+            chunks1.pop_back();
+            from = 1;
+            chunksLeft1--;
+            currentState_.preallocatedChunksLeft--;
+        } else if (chunks2.size() > 0) {
+            chunkid = chunks2.back();
+            chunks2.pop_back();
+            from = 2;
+            chunksLeft2--;
+            currentState_.preallocatedChunksLeft--;
+        }
+
+        return chunkid;
+    };
+
+    int from;
+    uint64_t chunkid;
+
+    if (!needClean) {
+        chunkid = smartPop(dirtyChunks_, cleanChunks_,
+            currentState_.dirtyChunksLeft, currentState_.cleanChunksLeft, from);
+        *isCleaned = (2 == from);
+        return chunkid;
+    }
+
+    *isCleaned = false;
+    chunkid = smartPop(cleanChunks_, dirtyChunks_,
+        currentState_.cleanChunksLeft, currentState_.dirtyChunksLeft, from);
+
+    if (1 == from || (2 == from && 0 == CleanChunk(chunkid, true))) {
+        *isCleaned = true;
+    }
+
+    return chunkid;
+}
+
+int FilePool::GetFile(const std::string& targetpath, char* metapage,
+    bool needClean, dynamicMetapage updateMetapageCb) {
     int ret = -1;
     int retry = 0;
+    bool isCleaned = false;
 
     while (retry < poolOpt_.retryTimes) {
         uint64_t chunkID;
         std::string srcpath;
         if (poolOpt_.getFileFromPool) {
-            std::unique_lock<std::mutex> lk(mtx_);
-            if (tmpChunkvec_.empty()) {
-                LOG(ERROR) << "no avaliable chunk!";
+            chunkID = GetChunk(needClean, &isCleaned);
```

Comment:
I'm not sure whether exist a chunk which id is 0, so it's better to let chunkID as an argument and check by the return value(true/false).

Reply:
Yeap, the valid chunk id start with 1.